app/src/main/java/org/oppia/android/app/home/WelcomeViewModel.kt

```diff
@@ -17,4 +17,15 @@ class WelcomeViewModel(
     fragment.requireContext(),
     oppiaClock
   ).getGreetingMessage()
+
+  override fun equals(other: Any?): Boolean {
+    if (this === other) {
+      return true
+    }
+    if (other == null || other.javaClass != javaClass) {
+      return false
+    }
+    val otherResult = other as WelcomeViewModel
+    return otherResult.profileName == this.profileName
```

Comment:
```suggestion
    return other is WelcomeViewModel && this.profileName == other.profileName
```

I think this can be simplified (note that the implicit casting is a nice simplification--checking the type of other lets you use it as that type since it's a val and cannot change). I removed the reference check since the content check is pretty simple, so the short-circuit doesn't add much.

Also, a few additional thoughts:
- I suggest adding a brief in-line comment explaining why equals() needs to be overridden since it's a bit unusual.
- We need to also override hashCode() to make sure that when two objects are equal, their hashes are also equal: https://docs.oracle.com/javase/7/docs/api/java/lang/Object.html#hashCode(). You can use ``Objects`` to do this more easily, e.g. something like: https://stackoverflow.com/a/52847604.
- Please add tests for each of the equals/hashcode methods (this may mean adding new view model-specific tests--that's fine; we want to make sure that equals & hashcode properly fit the expectations for Java objects: https://docs.oracle.com/javase/7/docs/api/java/lang/Object.html#equals(java.lang.Object)

Ditto on all of the above for the other new equals methods/modified view models.

Reply:
Done
I added tests based on the definitions for equals / hashcode (transitive, symmetric, etc.) but I don't know if I've set up the test activity in the best way since it now implements all the listener interfaces without using them. Not sure if there's actually a way around it, though. Let me know what you think!

Reply:
The test activity looks good--just had the few comments I left yesterday.